src/sentry/search/snuba/executors.py

```diff
@@ -0,0 +1,667 @@
+    TABLE_ALIAS = ""
+
+    @abstractmethod
+    def query(self):
+        raise NotImplementedError
+
+    def snuba_search(
+        self,
+        start,
+        end,
+        project_ids,
+        environment_ids,
+        sort_field,
+        cursor=None,
+        candidate_ids=None,
+        limit=None,
+        offset=0,
+        get_sample=False,
+        search_filters=None,
+    ):
+        """
+        This function doesn't strictly benefit from or require being pulled out of the main
+        query method above, but the query method is already large and this function at least
+        extracts most of the Snuba-specific logic.
+
+        Returns a tuple of:
+        * a sorted list of (group_id, group_score) tuples sorted descending by score,
+        * the count of total results (rows) available for this query.
+        """
+        if not self.QUERY_DATASET:
+            raise NotImplementedError
+
+        filters = {"project_id": project_ids}
+
+        if environment_ids is not None:
+            filters["environment"] = environment_ids
+
+        if candidate_ids:
+            filters[self.TABLE_ALIAS + "issue"] = sorted(candidate_ids)
+
+        conditions = []
+        having = []
+        for search_filter in search_filters:
+            if (
+                # Don't filter on issue fields here, they're not available
+                search_filter.key.name in self.issue_only_fields
+                or
+                # We special case date
+                search_filter.key.name == "date"
+            ):
+                continue
+            converted_filter = convert_search_filter_to_snuba_query(search_filter)
+            converted_filter = self.modify_converted_filter(
+                search_filter, converted_filter, environment_ids
+            )
+            field_name = self.TABLE_ALIAS + search_filter.key.name
+            # Ensure that no user-generated tags that clashes with aggregation_defs is added to having
+            if field_name in self.aggregation_defs and not search_filter.key.is_tag:
+                having.append(converted_filter)
+            else:
+                conditions.append(converted_filter)
+
+        extra_aggregations = self.dependency_aggregations.get(sort_field, [])
+        required_aggregations = set([sort_field, "total"] + extra_aggregations)
+        for h in having:
+            alias = h[0]
+            required_aggregations.add(alias)
+
+        aggregations = []
+        for alias in required_aggregations:
+            aggregations.append(self.aggregation_defs[alias] + [alias])
+
+        if cursor is not None:
+            having.append((sort_field, ">=" if cursor.is_prev else "<=", cursor.value))
+
+        selected_columns = []
+        if get_sample:
+            query_hash = md5(repr(conditions)).hexdigest()[:8]
+            selected_columns.append(
+                ("cityHash64", ("'{}'".format(query_hash), self.TABLE_ALIAS + "issue"), "sample")
```

Comment:
Is this the list of differences between snuba_search in the context of PostgresSnubaQueryExecutor and SnubaOnlyQueryExecutor ?
self.QUERY_DATASET
self.TABLE_ALIAS
self.issue_only_fields
self. modify_converted_filter 
self. aggregation_defs 
?

If yes, then I will suggest to implement a template method for this. https://en.wikipedia.org/wiki/Template_method_pattern
This is very similar with what you already have here, just add protected getter methods for those fields instead of accessing the fields directly, so the parent class cannot change the state of the subclasses.
The goal of this approach is that the parent class keeps control of the execution of snuba_query (which makes sense because the vast majority of the code is common) but it defines (through abstract method) some specific extension points the subclasses can customize. By avoiding sharing the state directly you guarantee that, in order to understand the subclass you do not have to check whether the parent class makes change to the state, thus reducing the coupling.